In [8]:
#import boto3

# Create a Boto3 S3 client
#s3 = boto3.client('s3')

# Define the bucket name and file name
#bucket_name = 'customer-churn-auto-insurance'
#file_name = 'churn.csv'

#Specify the local file path to save the downloaded file
#local_file_path = '/Users/Sarin/Desktop/Machine Learning/data.csv'

# Download the file from S3 to your local machine
#s3.download_file(bucket_name, file_name, local_file_path)

#print(f'{file_name} downloaded successfully to {local_file_path}')


In [9]:
#import pandas as pd

# Read the CSV file into a DataFrame
#df = pd.read_csv(local_file_path)

# Now you can work with the DataFrame as needed
#df


In [10]:
import pandas as pd

# Read the CSV file into a DataFrame
df = pd.read_csv(r"D:\Downloads\sula final project\data.csv")

# Now you can work with the DataFrame as needed
df

home_market_value        city  college_degree  good_credit  \
0           75000 - 99999    Burleson             0.0          1.0   
1                     NaN  Fort Worth             1.0          1.0   
2           50000 - 74999     Garland             1.0          1.0   
3           50000 - 74999     Bedford             1.0          1.0   
4         175000 - 199999      Forney             1.0          0.0   
...                   ...         ...             ...          ...   
1680904     50000 - 74999      Dallas             0.0          1.0   
1680905   175000 - 199999      Irving             0.0          0.0   
1680906   100000 - 124999  Fort Worth             0.0          1.0   
1680907     75000 - 99999  Fort Worth             0.0          1.0   
1680908   175000 - 199999      Desoto             0.0          1.0   

                        street_address state   latitude  individual_id  \
0                     3904 Brandy View    TX  32.533224   2.213023e+11   
1                      917 Snow Ridges    TX        NaN   2.213009e+11   
2           727 Amanda Avenue Apt. 680    TX  32.943144   2.213010e+11   
3           4526 Randall Land Apt. 065    TX  32.830994   2.213032e+11   
4          2673 Darren Rapid Suite 214    TX  32.788718   2.213029e+11   
...                                ...   ...        ...            ...   
1680904   383 Richards Forest Apt. 849    TX  32.912943   2.213026e+11   
1680905  02618 Nathan Harbor Suite 794    TX  32.824093   2.213004e+11   
1680906                 USCGC Browning    TX  32.753828   2.213033e+11   
1680907   9584 Pineda Corner Suite 736    TX        NaN   2.213024e+11   
1680908            49996 Jones Mission    TX  32.598562   2.213003e+11   

         length_of_residence  curr_ann_amt  ...  days_tenure  acct_suspd_date  \
0                        7.0   1038.346772  ...       2819.0              NaN   
1                        0.0    810.318741  ...        125.0              NaN   
2                       15.0   1115.395608  ...       4226.0              NaN   
3                        4.0    973.205184  ...       1650.0              NaN   
4                        0.0    781.841421  ...       6291.0              NaN   
...                      ...           ...  ...          ...              ...   
1680904                 15.0   1258.577836  ...       6291.0              NaN   
1680905                 11.0   1652.250350  ...       6291.0              NaN   
1680906                  4.0    715.997408  ...       1939.0              NaN   
1680907                 13.0    939.158494  ...       2663.0              NaN   
1680908                  1.0    880.656500  ...       4284.0              NaN   

         has_children  home_owner   county cust_orig_date  marital_status  \
0                 1.0         1.0  Johnson     2015-03-15         Married   
1                 1.0         0.0  Tarrant     2022-07-30         Married   
2                 1.0         1.0   Dallas     2011-05-08          Single   
3                 1.0         1.0  Tarrant     2018-05-27         Married   
4                 0.0         1.0  Kaufman     2005-09-11          Single   
...               ...         ...      ...            ...             ...   
1680904           0.0         1.0   Dallas     2005-09-11         Married   
1680905           0.0         1.0   Dallas     2005-09-11         Married   
1680906           0.0         1.0  Tarrant     2017-08-11          Single   
1680907           0.0         1.0  Tarrant     2015-08-18         Married   
1680908           0.0         1.0   Dallas     2011-03-11          Single   

         date_of_birth social_security_number age_in_years  
0           1985-09-21            394-XX-8902           37  
1           1970-11-25            886-XX-7817           52  
2           1947-03-02            667-XX-7169           76  
3           1974-11-24            374-XX-2595           48  
4           1960-08-28            460-XX-3202           62  
...           

In [11]:
# Retrieve column names and print
df.columns

Index(['home_market_value', 'city', 'college_degree', 'good_credit',
       'street_address', 'state', 'latitude', 'individual_id',
       'length_of_residence', 'curr_ann_amt', 'address_id', 'income', 'churn',
       'longitude', 'days_tenure', 'acct_suspd_date', 'has_children',
       'home_owner', 'county', 'cust_orig_date', 'marital_status',
       'date_of_birth', 'social_security_number', 'age_in_years'],
      dtype='object')

In [12]:
# Display information about the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1680909 entries, 0 to 1680908
Data columns (total 24 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   home_market_value       1588623 non-null  object 
 1   city                    1668842 non-null  object 
 2   college_degree          1680909 non-null  float64
 3   good_credit             1680909 non-null  float64
 4   street_address          1680909 non-null  object 
 5   state                   1680909 non-null  object 
 6   latitude                1427190 non-null  float64
 7   individual_id           1680909 non-null  float64
 8   length_of_residence     1680909 non-null  float64
 9   curr_ann_amt            1680909 non-null  float64
 10  address_id              1680909 non-null  float64
 11  income                  1680909 non-null  float64
 12  churn                   1680909 non-null  int64  
 13  longitude               1427190 non-null  float64
 14  da

In [13]:
# Check for missing values in each column
df.isnull().sum()

home_market_value           92286
city                        12067
college_degree                  0
good_credit                     0
street_address                  0
state                           0
latitude                   253719
individual_id                   0
length_of_residence             0
curr_ann_amt                    0
address_id                      0
income                          0
churn                           0
longitude                  253719
days_tenure                     0
acct_suspd_date           1487453
has_children                    0
home_owner                      0
county                      12067
cust_orig_date                  0
marital_status                  0
date_of_birth                   0
social_security_number          0
age_in_years                    0
dtype: int64

In [14]:
# Count the number of unique values in the 'churn' column
df['churn'].value_counts()

churn
0    1487453
1     193456
Name: count, dtype: int64

In [15]:
# Encode the 'marital_status' column using LabelEncoder
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['marital_status'] = le.fit_transform(df['marital_status'])

In [16]:
# Define the features to be used for prediction
feat_cols = ['curr_ann_amt', 'days_tenure', 'age_in_years', 'income', 'has_children',
       'length_of_residence', 'marital_status', 'home_owner', 'college_degree', 'good_credit']

# Extract features (X) and target variable (Y)
X = df[feat_cols]  # Features
Y = df['churn']     # Target variable

In [17]:
# Import train_test_split from sklearn.model_selection
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.33, random_state=42)

# Logistic Regression

In [18]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

# Scale the input features
scaler = StandardScaler()
x_train_scaled = scaler.fit_transform(x_train)

# Create a logistic regression model with increased max_iter
model = LogisticRegression(max_iter=1000)  # Increase max_iter as needed

# Train the model using the scaled training data
model.fit(x_train_scaled, y_train)


LogisticRegression(max_iter=1000)

In [19]:
import warnings
from sklearn.exceptions import ConvergenceWarning

# Suppress warnings about feature names and convergence
warnings.filterwarnings("ignore", category=UserWarning)
warnings.filterwarnings("ignore", category=ConvergenceWarning)

# Make predictions on the test set
y_pred1 = model.predict(x_test)

In [20]:
# Import accuracy_score from sklearn.metrics
from sklearn.metrics import accuracy_score

# Calculate the accuracy score
accuracy_score(y_pred1, y_test)

0.8846169100414638

# KNN Classifier

In [21]:
# Import K-Nearest Neighbors (KNN) classifier
from sklearn.neighbors import KNeighborsClassifier

In [22]:
# Create a K-Nearest Neighbors (KNN) classifier
knn = KNeighborsClassifier()

In [23]:
# Train the K-Nearest Neighbors (KNN) classifier using the training data
knn.fit(x_train, y_train)

KNeighborsClassifier()

In [24]:
# Make predictions on the test set using the trained K-Nearest Neighbors (KNN) classifier
y_pred3=knn.predict(x_test)

In [25]:
# Calculate the accuracy score of the K-Nearest Neighbors (KNN) classifier's predictions on the test set
accuracy_score(y_test,y_pred3)

0.881005949161709

# Decision Tree Classifier


In [26]:
# Import Decision Tree classifier
from sklearn.tree import DecisionTreeClassifier

In [27]:
# Create a Decision Tree classifier with a specified random state
model_dt=DecisionTreeClassifier(random_state = 48)

In [28]:
# Train the Decision Tree classifier using the training data
model_dt.fit(x_train,y_train)

DecisionTreeClassifier(random_state=48)

In [27]:
# Make predictions on the test set using the trained Decision Tree classifier
y_pred4=model_dt.predict(x_test)
y_pred4

array([1, 0, 0, ..., 0, 0, 0])

In [28]:
# Calculate the accuracy score of the Decision Tree classifier on the test set
model_dt.score(x_test,y_test)

0.8171462051559402

# Random Forest Classifier


In [29]:
# Import Random Forest classifier
from sklearn.ensemble import RandomForestClassifier

In [30]:
# Create a Random Forest classifier with 100 trees
model_rf=RandomForestClassifier(n_estimators=100)

In [ ]:
# Train the Random Forest classifier using the training data
model_rf.fit(x_train,y_train)

In [43]:
# Make predictions on the test set using the trained Random Forest classifier
y_pred5=model_rf.predict(x_test)

In [44]:
# Calculate the accuracy score of the Random Forest classifier on the test set
model_rf.score(x_test,y_test)

0.8828123309897242

## **Data Overview:**

- The dataset contains information on various attributes such as home market value, city, college degree, good credit, etc.
- It comprises 1,680,909 entries with 24 columns.
- Data types include float64, int64, and object.


## **Data Cleaning and Preprocessing:**

- Missing values were present in several columns, including 'home_market_value', 'city', 'latitude', 'longitude', 'acct_suspd_date', and 'county'.
- Categorical variables like 'marital_status' were encoded using LabelEncoder.
- Features and target variable were defined for model building.

## **Exploratory Data Analysis (EDA):**

- An analysis of missing values revealed that 'acct_suspd_date' had a significant number of missing values.
- The 'churn' column was imbalanced, with 1,487,453 entries for 'churn=0' and 193,456 entries for 'churn=1'.


## **Model Building and Evaluation:**

- Three machine learning models were trained and evaluated: Logistic Regression, K-Nearest Neighbors (KNN), and Decision Tree Classifier.
- Logistic Regression achieved an accuracy score of 0.885 on the test set.
- KNN Classifier achieved an accuracy score of 0.881 on the test set.
- Decision Tree Classifier achieved an accuracy score of 0.817 on the test set.
- A Random Forest Classifier was also trained, achieving an accuracy score of 0.883 on the test set.

## **Model Comparison:**

- Logistic Regression outperformed the other models in terms of accuracy.
- KNN Classifier and Random Forest Classifier also performed well but slightly lower than Logistic Regression.
- Decision Tree Classifier had the lowest accuracy among the models evaluated.
- Logistic Regression had the advantage of interpretability, while ensemble methods like Random Forest offered better predictive power.

## **Recommendations:**

- For this dataset and task, Logistic Regression may be the preferred choice due to its high accuracy and interpretability.
- However, if predictive power is prioritized and interpretability is less critical, Random Forest could be considered.
- Further optimization and feature engineering may improve the performance of the models, especially for Decision Tree Classifier.